Make custom elements's default ARIA id-ref work with shadow DOM

rniwa opened the pull request:
#### 8f927ad0d4312332b9c76c0b2219e02da04a202b
<pre>
Make custom elements&apos;s default ARIA id-ref work with shadow DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=245467">https://bugs.webkit.org/show_bug.cgi?id=245467</a>

Reviewed by Manuel Rego Casasnovas.

Make ElementInternals&apos;s id-ref ARIA attributes work with shadow DOM by recursively updating
relations between AX objects in AXObjectCache::updateRelationsIfNeeded.

* LayoutTests/accessibility/custom-elements/controls-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/controls-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/describedby-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/describedby-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/flowto-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/flowto-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/menuitem-shadow.html: Added.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateRelationsIfNeeded):
(WebCore::AXObjectCache::updateRelationsForTree): Extracted from updateRelationsIfNeeded.
* Source/WebCore/accessibility/AXObjectCache.h:

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible): Fixed a bug that this function returns false when
&quot;thisElement&quot; is disconnected but element is in the document tree.

Canonical link: <a href="https://commits.webkit.org/254762@main">https://commits.webkit.org/254762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0be5846bbc8c95d9ec057f270e68deea3a29ef29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99429 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156932 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33150 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82442 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/94270 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26362 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76932 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26251 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80988 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69251 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15068 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16011 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38981 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1426 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35095 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mrego reviewed (approved):
r=me

mrego commented on LayoutTests/platform/gtk/accessibility/custom-elements/describedby-shadow-expected.txt:
```diff
@@ -0,0 +1,32 @@
+FAIL accessibilityController.accessibleElementById("custom-1").detailsElements()[1].domIdentifier should be details2. Threw exception TypeError: undefined is not an object (evaluating 'accessibilityController.accessibleElementById("custom-1").detailsElements()[1]')
+FAIL accessibilityController.accessibleElementById("custom-1").errorMessageElements().length should be 2. Threw exception TypeError: undefined is not an object (evaluating 'accessibilityController.accessibleElementById("custom-1").errorMessageElements().length')
+FAIL accessibilityController.accessibleElementById("custom-1").errorMessageElements()[0].domIdentifier should be error1. Threw exception TypeError: undefined is not an object (evaluating 'accessibilityController.accessibleElementById("custom-1").errorMessageElements()[0]')
+FAIL accessibilityController.accessibleElementById("custom-1").errorMessageElements()[1].domIdentifier should be error2. Threw exception TypeError: undefined is not an object (evaluating 'accessibilityController.accessibleElementById("custom-1").errorMessageElements()[1]')
```
These shouldn't be needed after https://commits.webkit.org/254726@main, you can hopefully remove this WebKitGTK `-expected.txt` file.

rniwa committed:
Make custom elements's default ARIA id-ref work with shadow DOM
https://bugs.webkit.org/show_bug.cgi?id=245467

Reviewed by Manuel Rego Casasnovas.

Make ElementInternals's id-ref ARIA attributes work with shadow DOM by recursively updating
relations between AX objects in AXObjectCache::updateRelationsIfNeeded.

* LayoutTests/accessibility/custom-elements/controls-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/controls-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/describedby-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/describedby-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/flowto-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/flowto-shadow.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-shadow-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/menuitem-shadow.html: Added.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateRelationsIfNeeded):
(WebCore::AXObjectCache::updateRelationsForTree): Extracted from updateRelationsIfNeeded.
* Source/WebCore/accessibility/AXObjectCache.h:

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible): Fixed a bug that this function returns false when
"thisElement" is disconnected but element is in the document tree.

Canonical link: https://commits.webkit.org/254762@main
